examples/car_fsm.erl

```diff
@@ -0,0 +1,543 @@
+%% API
+%% -----------------------------------------------------------------------------
+
+
+start_link() ->
+  gen_statem:start_link({local, ?NAME}, ?MODULE, [], []).
+
+stop() ->
+  gen_statem:stop(?NAME).
+
+accelerate(Value) ->
+  gen_statem:call(?NAME, {accelerate, Value}).
+
+brake(Value) ->
+  gen_statem:call(?NAME, {brake, Value}).
+
+travel(Distance) ->
+  gen_statem:call(?NAME, {travel, Distance}).
+
+refuel(Amount) ->
+  gen_statem:call(?NAME, {refuel, Amount}).
+
+
+%% -----------------------------------------------------------------------------
+%% gen_statem callbacks
+%% -----------------------------------------------------------------------------
+
+
+init([]) ->
+  {ok, stopped, #gen_state{fuel = ?MAX_FUEL, speed = 0}}.
+
+callback_mode() ->
+  state_functions.
+
+stopped({call, From}, {accelerate, Value}, S) ->
+  accelerate_helper(From, Value, S);
+stopped({call, From}, {refuel, Value}, S) ->
+  refuel_helper(From, Value, S).
+
+slow({call, From}, {accelerate, Value}, S) ->
+  accelerate_helper(From, Value, S);
+slow({call, From}, {brake, Value}, S) ->
+  accelerate_helper(From, -Value, S);
+slow({call, From}, {travel, Value}, S) ->
+  travel_helper(From, Value, S);
+slow({call, From}, {refuel, Value}, S) ->
+  refuel_helper(From, Value, S).
+
+normal({call, From}, {accelerate, Value}, S) ->
+  accelerate_helper(From, Value, S);
+normal({call, From}, {brake, Value}, S) ->
+  accelerate_helper(From, -Value, S);
+normal({call, From}, {travel, Value}, S) ->
+  travel_helper(From, Value, S);
+normal({call, From}, {refuel, Value}, S) ->
+  refuel_helper(From, Value, S).
+
+fast({call, From}, {accelerate, Value}, S) ->
+  accelerate_helper(From, Value, S);
+fast({call, From}, {brake, Value}, S) ->
+  accelerate_helper(From, -Value, S);
+fast({call, From}, {travel, Value}, S) ->
+  travel_helper(From, Value, S);
+fast({call, From}, {refuel, Value}, S) ->
+  refuel_helper(From, Value, S).
+
+too_fast({call, From}, {accelerate, Value}, S) ->
+  accelerate_helper(From, Value, S);
+too_fast({call, From}, {brake, Value}, S) ->
+  accelerate_helper(From, -Value, S);
+too_fast({call, From}, {travel, Value}, S) ->
+  travel_helper(From, Value, S);
+too_fast({call, From}, {refuel, Value}, S) ->
+  refuel_helper(From, Value, S).
+
+terminate(_Reason, _StateName, _State) ->
+  ok.
+
+code_change(_OldVsn, StateName, State, _Extra) ->
+  {ok, StateName, State}.
+
+
+%% -----------------------------------------------------------------------------
+%% gen_statem helpers
+%% -----------------------------------------------------------------------------
+
+
+accelerate_helper(From, Value, S) ->
+  #gen_state{fuel = Fuel, speed = Speed} = S,
+  {Distance, Acceleration, Burnt} =
+    acceleration_calculations({Speed, Value}, Fuel),
+  StateName = state_name(Speed + Acceleration),
+  gen_statem:reply(From, {Distance, Burnt}),
+  {next_state, StateName, S#gen_state{fuel = Fuel - Burnt,
```

Comment:
I think these two lines look better as
```erl
  NS = S#gen_state{fuel = Fuel - Burnt, speed = Speed + Acceleration},
  {next_state, StateName, NS}.
```
(similarly in other places in this module).